src/EventStore.Core/Services/Transport/Enumerators/Enumerator.AllSubscriptionFiltered.cs

```diff
@@ -71,462 +59,380 @@ public class AllSubscriptionFiltered : IAsyncEnumerator<ReadResponse> {
+				await foreach (var liveEvent in _liveEvents.Reader.ReadAllAsync(ct)) {
+					var sequenceCorrect = liveEvent.SequenceNumber == sequenceNumber + 1;
+					sequenceNumber = liveEvent.SequenceNumber;
 
-								await _channel.Writer.WriteAsync(new ReadResponse.CheckpointReceived(
-										commitPosition: position.CommitPosition,
-										preparePosition: position.PreparePosition), ct);
-							}
+					if (liveEvent.ResolvedEvent.HasValue && liveEvent.ResolvedEvent.Value.OriginalPosition <= checkpoint) {
+						// skip because this event has already been sent towards the client
+						// (or the client specified a checkpoint for events that don't exist yet and so
+						// is not interested in them)
+						// nb: we can skip this event even if it had the wrong sequence number, it means
+						// we would have skipped the events that got discarded anyway.
+						continue;
+					}
 
-							ReadPage(nextPosition, OnMessage);
-							return;
+					if (liveEvent.Checkpoint.HasValue && liveEvent.Checkpoint.Value < checkpoint) {
+						// skip because the checkpoint received is earlier than the last event or checkpoint sent towards the client
+						continue;
+					}
 
-						case FilteredReadAllResult.Expired:
-							ReadPage(
-								Position.FromInt64(
-									completed.CurrentPos.CommitPosition,
-									completed.CurrentPos.PreparePosition),
-								OnMessage);
-							return;
+					if (!sequenceCorrect) {
+						// there's a gap in the sequence numbers, at least one live event was discarded
+						// due to the live channel becoming full.
+						// switch back to catchup to make sure we didn't miss anything we wanted to send.
+						await NotifyFellBehind(checkpoint, ct);
 
-						case FilteredReadAllResult.AccessDenied:
-							Fail(new ReadResponseException.AccessDenied());
-							return;
+						// issue a checkpoint before falling behind to catch-up to make sure that
+						// at least one checkpoint is issued within the checkpoint interval
+						await SendCheckpointToSubscription(checkpoint, ct);
 
-						case FilteredReadAllResult.InvalidPosition:
-							Fail(new ReadResponseException.InvalidPosition());
-							return;
+						return (checkpoint, sequenceNumber);
+					}
 
-						default:
-							Fail(ReadResponseException.UnknownError.Create(completed.Result));
-							return;
+					if (liveEvent.ResolvedEvent.HasValue) {
+						// this is the next event to send towards the client. send it and update `checkpoint`
+						await SendEventToSubscription(liveEvent.ResolvedEvent.Value, ct);
+						checkpoint = liveEvent.ResolvedEvent.Value.OriginalPosition!.Value;
+					} else if (liveEvent.Checkpoint.HasValue) {
+						// this is a valid checkpoint that we can send towards the client. send it and update `checkpoint`
+						await SendCheckpointToSubscription(liveEvent.Checkpoint.Value, ct);
+						checkpoint = new TFPos(liveEvent.Checkpoint.Value.CommitPosition, liveEvent.Checkpoint.Value.PreparePosition);
```

Comment:
```suggestion
						checkpoint = liveEvent.Checkpoint.Value;
```
any reason not to do this?

Reply:
thanks, not sure why i wrote it that way :smile: